clang/lib/CodeGen/Targets/NVPTX.cpp

```diff
@@ -302,6 +299,19 @@ void NVPTXTargetCodeGenInfo::addNVVMMetadata(
       llvm::ConstantAsMetadata::get(GV), llvm::MDString::get(Ctx, Name),
       llvm::ConstantAsMetadata::get(
           llvm::ConstantInt::get(llvm::Type::getInt32Ty(Ctx), Operand))};
+  // Append metadata to nvvm.annotations
+  MD->addOperand(llvm::MDNode::get(Ctx, MDVals));
+}
+
+void NVPTXTargetCodeGenInfo::addNVVMGridConstantMetadata(
+    llvm::GlobalValue *GV, const SmallVectorImpl<int> &GridConstantArgs) {
+  llvm::Module *M = GV->getParent();
+  llvm::LLVMContext &Ctx = M->getContext();
+
+  // Get "nvvm.annotations" metadata node
+  llvm::NamedMDNode *MD = M->getOrInsertNamedMetadata("nvvm.annotations");
```

Comment:
Strictly speaking, `grid_constant` info does not have to be `nvvm.annotation` any more. We could convert it to a parameter attribute.

Reply:
Yea, I completely agree, I think almost all the other nvvm.annotations can be converted as well. For this MR I want to lay down the framework for a couple and once that is setup it should be fairly trivial to convert all the others.

Specifically for grid_constant can we just upgrade to the existing `readonly` parameter attribute?